### ISSUE TITLE
the file whose Directory address is 'tools/llm/llm.py' 

### ISSUE THREAD

**Issue by Thovenfish** [jack-and-rozz/vocabulary_adaptation#2]
Thanks for your great work! 
​I have some questions about the code which you uploaded on github.

-  where is the file whose Directory address is 'tools/llm/llm.py' ?  

- I also wonder whether these codes are complete. Can I run them and get the results on them directly?

Thanks again.

**Comment by jack-and-rozz** [comment]
Thank you for your attention to our work!

[About the LLM tool]
Only `map_embeddings.sh` uses `tools/llm/llm.py`, which is namely a script to map embeddings from a target domain (CBoW space) to a source domain (NMT-model space).  Does this answer your question?

[Completeness of these codes]
I'm not sure what you meant by "directly", but this code is almost equivalent to the repository used for our experiments (i.e., you can train and evaluate models using this code and the tools specified in the README). 
More specifically, 
- Basically, what you need to do is 1) download tools following the README,  2) manually download data and place them on the data directory,  3) run `train.sh` with the identifier of the model you want to train. Although this repository contains several scripts, most of them are automatically run in `train.sh`. 
- As mentioned above, the process of downloading data is not automated. This is because some data requires a request and approval for downloading.
- As this is not a library but code for experiments, we cannot guarantee that this code works in any environments and you may have a small error depending on your environment or situation. 
- A quick glance at `train.sh` and `const.sh` may give you an overview of this repository.

If you have another detailed question about this code, feel free to ask here.

**Comment by Thovenfish** [comment]
Thanks for your reply. 
I have trained the out-domian model by my codes and I have some in-domian datasets. Now I want to use llm.py to map the embedding. What I should do is that:
- train the cbow model of in-domian;
- and then use llm.py to map the embeddings.

After reading your codes, I have some other questions as follows:
1. When training the out-domian model, Is the dictionary of the out-domian model generated by word2vec before fairseq-preprocess?
2. When mapping embeddings, the out-domian token embeddings which are generated by fairseq have extra 4 special tokens('pad'、'/s'、'unk'、's') .  What I understand is that these special tokens don't participate in the mapping and are directly applied to the corresponding tokens('pad'、'/s'、'unk'、's' in in-domain models training) of the in-domain model. I am not sure what I understand is right.

Hope I have described my problem clearly.😀
Thank you sincerely.

PS: In the readme of ./tools/llm/:
```
 $ python -m llm [SRC_EMB] [TRG_EMB] --num-neighbors 10 < [INPUT] 
``` 
I find that the position of [SRC_EMB] and [TRG_EMB] should probably be swapped?

**Comment by jack-and-rozz** [comment]
> 1. When training the out-domian model, Is the dictionary of the out-domian model generated by word2vec before fairseq-preprocess?

Yes, although there could be several options, I also trained CBoW embeddings from the out-domain data and picked up top-15996 frequent subwords as a dictionary from the output file of word2vec. This is because I didn't want to make the process of building vocabularies heterogeneous and wanted to define the vocabulary of a domain before running fairseq-preprocess.

> 2. When mapping embeddings, the out-domain token embeddings which are generated by fairseq have extra 4 special tokens... 

Yes, your understanding is correct.  The embeddings of the four special tokens are not overwritten before fine-tuning. The in-domain model to be trained just inherits the four embeddings from the out-domain model (`fairseq/trainer.py` shows how the embeddings are overwritten and the modification from the original fairseq code is very small).

[About the README of the LLM repository]
Sorry for the misleading descriptions in my scripts.
I guess both the README and `map_embedding.sh` are correct because we tried to map embeddings from an **in-domain** space to an **out-domain** space.  The README of LLM denotes the departure and the destination of **mapping** as "source" and "target", respectively. On the other hand, in my scripts, I used the words "source" and "target" as "source domain (out-domain)" and "target domain (in-domain)". So the direction of DA is source-to-target but the direction of the projection is target-to-source in our situation. Does this explanation make sense?

**Comment by Thovenfish** [comment]
Thanks for your clear explanation. It helps me a lot. Thanks again ~

**Comment by Thovenfish** [comment]
I am so sorry. I have another question now. When using monolingual data, is the vocabulary of in-domain model generated by  in-domain parallel datasets and monolingual datasets？

**Comment by jack-and-rozz** [comment]
Yes, we built the vocabulary from both datasets since we considered that it would be natural to use available data as far as possible in practical situations. The same is true for CBoW training.

Just in case, note that the monolingual data was synthetically generated by separating a parallel corpus as the paper describes. 
Concretely,  we first divided `N` numbers of parallel data into `X` parallel data and `N-X` monolingual data. And then, we used the first half and the second half of the `N-X` data as the monolingual data for the source and the target language, respectively.  

FYI, the current README only displays the model names without monolingual data. 
When using monolingual data, the identifier to specify the amount of data for building vocabulary and training CBoW needs to be changed. For example, from `jesc_sp16000@aspec_sp16000.va.v_aspec_sp16000_100k.llm-idt.nn10.100k` to
`jesc_sp16000@aspec_sp16000.va.v_aspec_sp16000_100kmono.llm-idt.nn10.100k`.
